File: sol/2528/2528.py
class Solution:
    def maxPower(self, stations: List[int], r: int, k: int) -> int:
        n = len(stations)
        arr = [0] * (n + 1)
        for i, v in enumerate(stations):
            arr[max(0, i - r)] += v
            arr[min(n, i + r + 1)] -= v
        s = min(stations)
        e = sum(stations) + k + 1
        while s < e:
            m = (s + e) // 2
            feasible = True
            val = 0
            remain = k
            diff = arr.copy()
            for i in range(n):
                val += diff[i]
                if val >= m:
                    continue
                x = m - val
                if x > remain:
                    feasible = False
                    break
                else:
                    val += x
                    remain -= x
                    diff[min(n, i + 2 * r + 1)] -= x
            if feasible:
                s = m + 1
            else:
                e = m
        return e - 1

File: sol/2528/2528.cpp
struct Solution {
    using ll = long long;
    long long maxPower(vector<int>& stations, int r, int k) {
        int n = stations.size();
        vector<ll> arr(n + 1, 0);
        ll s = -1, e = k + 1;
        for (int i = 0; i < n; ++i){
            int v = stations[i];
            arr[max(0, i - r)] += v;
            arr[min(n, i + r + 1)] -= v;
            if (s == -1 || s > v) {
                s = v;
            }
            e += v;
        }
        while (s < e){
            ll m = s + (e - s) / 2, val = 0, remain = k;
            bool feasible = true;
            vector<ll> diff{arr};
            for (int i = 0; i < n; ++i){
                val += diff[i];
                if (val >= m){
                    continue;
                }
                ll x = m - val;
                if (x > remain){
                    feasible = false;
                    break;
                } else {
                    val += x;
                    remain -= x;
                    diff[min(n, i + r * 2 + 1)] -= x;
                }
            }
            if (feasible){
                s = m + 1;
            } else {
                e = m;
            }
        }
        return e - 1;
    }
};
